crates/mysten-network/src/server.rs

```diff
@@ -205,22 +211,132 @@ impl<M: MetricsCallbackProvider> ServerBuilder<M> {
     }
 }
 
-async fn tcp_listener_and_update_multiaddr<T: ToSocketAddrs>(
+async fn listen_and_update_multiaddr<T: ToSocketAddrs>(
     address: &Multiaddr,
     socket_addr: T,
-) -> Result<(Multiaddr, TcpListenerStream)> {
-    let (local_addr, incoming) = tcp_listener(socket_addr).await?;
+    tls_config: Option<ServerConfig>,
+) -> Result<(
+    Multiaddr,
+    impl Stream<Item = std::io::Result<TcpOrTlsStream>>,
+)> {
+    let listener = TcpListener::bind(socket_addr).await?;
+    let local_addr = listener.local_addr()?;
     let local_addr = update_tcp_port_in_multiaddr(address, local_addr.port());
-    Ok((local_addr, incoming))
+
+    let tls_acceptor = tls_config.map(|tls_config| TlsAcceptor::from(Arc::new(tls_config)));
+    let incoming = TcpOrTlsListener::new(listener, tls_acceptor);
+    let stream = async_stream::stream! {
+        loop {
+            yield incoming.accept().await;
+        }
+    };
+
+    Ok((local_addr, stream))
 }
 
-async fn tcp_listener<T: ToSocketAddrs>(address: T) -> Result<(SocketAddr, TcpListenerStream)> {
-    let listener = TcpListener::bind(address).await?;
-    let local_addr = listener.local_addr()?;
-    let incoming = TcpListenerStream::new(listener);
-    Ok((local_addr, incoming))
+pub struct TcpOrTlsListener {
+    listener: TcpListener,
+    tls_acceptor: Option<TlsAcceptor>,
 }
 
+impl TcpOrTlsListener {
+    fn new(listener: TcpListener, tls_acceptor: Option<TlsAcceptor>) -> Self {
+        Self {
+            listener,
+            tls_acceptor,
+        }
+    }
+
+    async fn accept(&self) -> std::io::Result<TcpOrTlsStream> {
+        let (stream, addr) = self.listener.accept().await?;
+        if self.tls_acceptor.is_none() {
+            return Ok(TcpOrTlsStream::Tcp(stream, addr));
+        }
+
+        // Determine whether new connection is TLS.
```

Comment:
Is there a test for having both TLS and non-TLS connections to a server? And is the detection logic expected to be a temporary measure?

Reply:
Agree that we need an explicit test checking that both TLS and not-TLS connections work simultaneously.

As for if this is a short-term or long-term measure, at the very least this is required in the short term. In the long term it could likely be beneficial to still allow unsecured connections in some scenarios, if only for convenience, but we should definitely move everything we possibly can to use secured connections.

Reply:
My proposal would be to remove client-side support for TCP connections once this is released to all validators, but we may want to keep the server-side support around for longer depending on how frequently we demand that clients update their binaries

As for the tests, I manually ran simtests in both modes by uncommenting the currently-disabled client code. I think the simultaneous-modes test should wait until enabling TLS-by-default code on the client side in a future PR